### PR TITLE
Photon Controller Driver v1.0: Added functionality for create docker-machine with ssh enabled.

### PR DESCRIPTION
**Pull request by faijaz**
1) Driver create functionality is implemented for Docker-Machine to run on Photon Controller.
2) Added Photon OS provisioner to make Photon OS as one of the options compatible for docker.

### CODE REVIEW

File: drivers/photoncontroller/photoncontroller.go [hunk 1]
```diff
@@ -9,14 +9,36 @@ package photoncontroller
...
 
 	"github.com/docker/machine/libmachine/drivers"
 	"github.com/docker/machine/libmachine/mcnflag"
 	"github.com/docker/machine/libmachine/state"
+	"github.com/vmware/photon-controller-go-sdk/photon"
+)
+
+const (
+	driverName = "photoncontroller"
+	defaultEndpoint = "http://10.118.97.200:9000"
```
Comment (AlainRoy) [inline drivers/photoncontroller/photoncontroller.go]: This default endpoint looks suspiciously like your devbox's IP address. Is it? 

Do we need a default here? I'd recommend removing it (if it's not needed) or setting it to something that has to be changed. I'd recommend an IP from the "documentation range", like "192.0.2.2". 

https://tools.ietf.org/html/rfc5737

File: drivers/photoncontroller/photoncontroller.go [hunk 2]
```diff
@@ -9,14 +9,36 @@ package photoncontroller
...
+	defaultAuthEndpoint = "10.146.64.236"
+	defaultSSHUser = "root"
+	defaultSSHKeyPath = "~/.ssh/id_rsa"
```
Comment (AlainRoy) [inline drivers/photoncontroller/photoncontroller.go]: Does this work? Normally ~ is only expanded by shells.

File: drivers/photoncontroller/photoncontroller.go [hunk 3]
```diff
@@ -27,63 +49,230 @@ func (e NotLoadable) Error() string {
 	return fmt.Sprintf("Driver %q not found. Do you have the plugin binary accessible in your PATH?", e.Name)
 }
 
-func NewDriver(Name string) drivers.Driver {
+func NewDriver(hostName, storePath string) *Driver {
 	return &Driver{
-		Name: Name,
+		Tenant:         "TenantsName",
+		Project:        "626ce36a-8d61-4b69-8880-81829aac7b8c",
```
Comment (AlainRoy) [inline drivers/photoncontroller/photoncontroller.go]: Is this hard-coded to your project? (And same for the image below? Does it matter, or is this just an example?

If it's just an example, I'd recommend adding a comment to make that clear. If it's not just an example, can you comment on the plans for making this more generic in the future?

File: drivers/photoncontroller/photoncontroller.go [hunk 4]
```diff
@@ -27,63 +49,230 @@ func (e NotLoadable) Error() string {
...
 func (d *Driver) GetURL() (string, error) {
-	return "", NotLoadable{d.Name}
+	ip, err := d.GetIP()
+	if err != nil {
+		return "", err
+	}
+	if ip == "" {
+		return "", nil
+	}
+	return fmt.Sprintf("tcp://%s", net.JoinHostPort(ip, "2376")), nil
```
Comment (AlainRoy) [inline drivers/photoncontroller/photoncontroller.go]: Can you use DefaultPort from libmachine/engine/engine.go?

File: drivers/photoncontroller/photoncontroller.go [hunk 5]
```diff
@@ -27,63 +49,230 @@ func (e NotLoadable) Error() string {
...
+	}
+
+	// Waiting for create VM task completion
+	vmCreateTask, err = d.getClient().Tasks.Wait(vmCreateTask.ID)
+	if (err != nil) {
+		return fmt.Errorf("Error creating VM: ", err)
+	}
+
+	d.VMId = vmCreateTask.Entity.ID
+	fmt.Println("VM is created with Id: ", d.VMId)
```
Comment (AlainRoy) [inline drivers/photoncontroller/photoncontroller.go]: nit: is -> was

File: drivers/photoncontroller/photoncontroller.go [hunk 6]
```diff
@@ -27,63 +49,230 @@ func (e NotLoadable) Error() string {
...
+				ipAddress = val.(string)
+				break
+			}
+		}
+
+		if ipAddress != "" {
+			break;
+		}
+		numRetries++
+		time.Sleep(10000 * time.Millisecond)
```
Comment (AlainRoy) [inline drivers/photoncontroller/photoncontroller.go]: I suggest that you poll more quickly, like once a second. Users will judge us based on their experience here, and you might wait nearly 10 seconds longer than you need to. Polling once a second shouldn't cause too much load. 

In Java, we could write this more clearly by doing: 
    TimeUnit.SECONDS.toMicros(10);
Is there an equivalent in Go?

File: drivers/photoncontroller/photoncontroller.go [hunk 7]
```diff
@@ -27,63 +49,230 @@ func (e NotLoadable) Error() string {
...
+	}
+
+	d.IPAddress = ipAddress
```
Comment (AlainRoy) [inline drivers/photoncontroller/photoncontroller.go]: You print lots of things out: is it worth printing out the IP address here?

File: libmachine/provision/photon.go [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+
+type PhotonProvisioner struct {
+	SystemdProvisioner
+}
+
+func (provisioner *PhotonProvisioner) String() string {
+	return "photon"
+}
+
+func (provisioner *PhotonProvisioner) CompatibleWithHost() bool {
```
Comment (AlainRoy) [inline libmachine/provision/photon.go]: This function looks like it's worth describing in a short comment. In general, I'd appreciate more comments in the code.

File: libmachine/provision/photon.go [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+		hostname,
+		hostname,
+	)); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (provisioner *PhotonProvisioner) Package(name string, action pkgaction.PackageAction) error {
```
Comment (AlainRoy) [inline libmachine/provision/photon.go]: A comment here would be good. Note that we are using Photon's tdnf

File: libmachine/provision/photon.go [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+	}
+
+	if err := provisioner.Service("docker", serviceaction.Enable); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func installPackages(provisioner *PhotonProvisioner) error {
```
Comment (AlainRoy) [inline libmachine/provision/photon.go]: This definitely needs a comment. I don't know what these packages include, nor why you need them.

File: libmachine/provision/photon.go [hunk 11]
```diff
@@ -0,0 +1,180 @@
...
+	swarmOptions.Env = engineOptions.Env
+
+	provisioner.EngineOptions.StorageDriver = "overlay"
+
+	log.Info("Installing packages")
+	if err := installPackages(provisioner); err != nil {
+		return err
+	}
+
+	log.Info("setting hostname")
```
Comment (AlainRoy) [inline libmachine/provision/photon.go]: You have inconsistent capitalization in the logging here:

Installing packages vs setting hostname

File: vendor/github.com/vmware/photon-controller-go-sdk/photon/apitypes.go [hunk 12]
```diff
@@ -0,0 +1,598 @@
+// Copyright (c) 2016 VMware, Inc. All Rights Reserved.
+//
+// This product is licensed to you under the Apache License, Version 2.0 (the "License").
+// You may not use this product except in compliance with the License.
+//
+// This product may include a number of subcomponents with separate copyright notices and
+// license terms. Your use of these subcomponents is subject to the terms and conditions
+// of the subcomponent's license, as noted in the LICENSE file.
+
+package photon
```
Comment (AlainRoy) [inline vendor/github.com/vmware/photon-controller-go-sdk/photon/apitypes.go]: I'm not a go expert: I think that godeps added the SDK--is that accurate?

Do we need any internal docs on how to update to a new SDK, or what version of the SDK we used?

**Comment by frapposelli** [comment]
Please note that the Photon provisioner will not be accepted until we have official docker packages, see https://github.com/docker/machine/pull/2344 for more context.

File: drivers/photoncontroller/photoncontroller.go (hunk 3 above)
Comment (faijaz) [inline drivers/photoncontroller/photoncontroller.go]: This is just an example. We will read the values as part of env variables or command line. I will add some comments here.

File: vendor/github.com/vmware/photon-controller-go-sdk/photon/apitypes.go (hunk 12 above)
Comment (faijaz) [inline vendor/github.com/vmware/photon-controller-go-sdk/photon/apitypes.go]: My code will be compatible with the SDK that I have added here using godeps. The general guideline for updating any packages is part of docker machine contribution. That will be applicable for Photon Controller as well. I haven't done anything special here for SDK.

File: drivers/photoncontroller/photoncontroller.go (hunk 1 above)
Comment (faijaz) [inline drivers/photoncontroller/photoncontroller.go]: Done.

File: drivers/photoncontroller/photoncontroller.go (hunk 2 above)
Comment (faijaz) [inline drivers/photoncontroller/photoncontroller.go]: Not actually. But I didn't want to make it specific to my user. I have changed it now to some generic path.

File: drivers/photoncontroller/photoncontroller.go (hunk 4 above)
Comment (faijaz) [inline drivers/photoncontroller/photoncontroller.go]: Done.

File: drivers/photoncontroller/photoncontroller.go (hunk 5 above)
Comment (faijaz) [inline drivers/photoncontroller/photoncontroller.go]: Done.

File: drivers/photoncontroller/photoncontroller.go (hunk 6 above)
Comment (faijaz) [inline drivers/photoncontroller/photoncontroller.go]: Changed it to check every second.

File: drivers/photoncontroller/photoncontroller.go (hunk 7 above)
Comment (faijaz) [inline drivers/photoncontroller/photoncontroller.go]: Yes, add the code to print IP as well.

File: libmachine/provision/photon.go (hunk 8 above)
Comment (faijaz) [inline libmachine/provision/photon.go]: Added comments all over the file to explain the functions that provisioner implements.

File: libmachine/provision/photon.go (hunk 9 above)
Comment (faijaz) [inline libmachine/provision/photon.go]: Added comment for function and related to tdnf usage as well.

File: libmachine/provision/photon.go (hunk 10 above)
Comment (faijaz) [inline libmachine/provision/photon.go]: Add details in comments.

File: libmachine/provision/photon.go (hunk 11 above)
Comment (faijaz) [inline libmachine/provision/photon.go]: Fixed it.

File: drivers/photoncontroller/photoncontroller.go [hunk 13]
```diff
@@ -27,63 +50,228 @@ func (e NotLoadable) Error() string {
...
-func NewDriver(Name string) drivers.Driver {
+func NewDriver(hostName, storePath string) *Driver {
 	return &Driver{
-		Name: Name,
+		Tenant:         "TenantsName",
+		Project:        "626ce36a-8d61-4b69-8880-81829aac7b8c",
+		VMFlavor:       "VMFlavor",
+		DiskFlavor:     "DiskFlavor",
+		DiskName:       "disk-10",
+		Image:          "707f566f-c3d0-44b3-8eb6-e66a087f46dc",
```
Comment (AlainRoy) [inline drivers/photoncontroller/photoncontroller.go]: Okay for now, but would still be good to add a comment here in the future. Or maybe even a clearly bogus unique ID, like 000000-0000...

**Comment by AlainRoy** [comment]
Looks good to me: I think you should merge it.

**Pull request merged**